enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,294 @@
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+MCO container runtime config controller can add unit tests and e2e tests.
+- unit test: verify the policies from ImagePolicy instances merged correctly and the controller writes correct format policy.json
+- e2e test: verify the MCO writes configuration to the correct location
+
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+#### Removing a deprecated feature
+
+Not applicable.
+
+### Upgrade / Downgrade Strategy
+
+Upgrade expectations:
+In order to use ImagePolicy, an existing cluster will have to make on upgrade.
+
+Downgrade expectations: 
+- Not applicable. 
+  If `N` does not support ImagePolicy CRD, `N+1`supports IamgePolicy. It is not expected that the ImagePolicy related to the failure `N->N+1`. New resources should not be created during `N->N+1`.
+
+
+### Version Skew Strategy
+
+The MCO does not require skew check.
+
+### Operational Aspects of API Extensions
+
+- Use-cases that requires over 1000 instances of the CRD is rare, not impacting
+    general API throughput.
+
+#### Failure Modes
+
+- CR field syntax error failure reported from CLI
+- CR conflicting value failure reported by machine config controller logs
+- MCO rolling out configuration file failure reported by MCO
+- If the signature validation fails, then CRI-O will report that in the same way as for any other pull failure. Further enhancements to the kubelet, CRI and CRI-O error handling can be achieved in future Kubernetes releases.
+
+The above errors should not impact the overall cluster health
+OCP Node team is likely to be called upon in case of escalation with one of the failure modes.
+
+#### Support Procedures
+
+- detect the failure modes in a support situation, describe possible symptoms (events, metrics,
+  alerts, which log output in which component)
+  - If the ImagePolicy CR does not follow the syntax requirements, api-server will fail when creating the objects.
+  - If the ImagePolicy CR is not rolled out by MCO, machine-config-controller logs will shows error with the reason.
+
```

Comment:
If the policy is too strict for an image and it fails to pull, will logs show which policy was violated and... then folks have to know about the ImagePolicy CRD out of band to be able to map that CRI-O complaint back up to the conflicting ImagePolicy, if they decide to address the conflict by adjusting policy instead of by adjusting the image they're pulling?

Also in this bucket are network-config issues that block the cluster from being able to access the signature store.  Are there proxy/restricted-network considerations for either implementation or debug-ability?

Reply:
> Are there proxy/restricted-network considerations for either implementation or debug-ability?

The signatures are stored in the same registry repos as the images themselves, so this does not introduce a new point of failure.

There is _some_ potential for the registry reporting “image not found” on the signature fake-image, and that would show up for the user as “signature missing” without directly pointing at the registry as a culprit.

---

A user-visible process change for disconnected users is the need to mirror signatures together with the application images. I don’t know which of the `oc` mirroring subcommands support that.